pkg/handler/collector/rekor/rekor.go

```diff
@@ -0,0 +1,331 @@
+// limitations under the License.
+
+package rekor
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/gob"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io/fs"
+	"log"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/sigstore/rekor-monitor/mirroring"
+	"github.com/sigstore/rekor/pkg/client"
+	c "github.com/sigstore/rekor/pkg/generated/client"
+	"github.com/sigstore/rekor/pkg/util"
+)
+
+type rekorCollector struct {
+	server string
+	client *c.Rekor
+	// lastDownload time.Time
+	poll     bool
+	interval time.Duration
+}
+
+func NewRekorCollector(rekorServerURL string, poll bool, interval time.Duration) (*rekorCollector, error) {
+	// Get the Rekor client.
+	r, err := getRekorClient(PublicRekorServerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build a RekorCollector.
+	rc := &rekorCollector{
+		server:   rekorServerURL,
+		client:   r,
+		poll:     poll,
+		interval: interval,
+	}
+	return rc, nil
+}
+
+const (
+	CollectorRekor       = "rekor"
+	PublicRekorServerURL = "https://rekor.sigstore.dev"
+	logInfoFileName      = "logInfo.txt"
+)
+
+func Type() string {
+	return CollectorRekor
+}
+
+func (rc *rekorCollector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
+	// Get all the entries since last checkpoint.
+	// If there is no checkpoint, get all entries.
+
+	logInfoFile := "logInfo.txt"
+	sth, first, err := rc.retrieveLastCheckpoint(logInfoFile)
+	if err != nil {
+		return err
+	}
+
+	pubkey, err := mirroring.GetPublicKey(rc.client)
+	if err != nil {
+		return fmt.Errorf("getting public key: %v", err)
+	}
+
+	// TODO: Verify using public key from TUF
+	// Verify the checkpoint with the server's public key
+	if err := mirroring.VerifySignedTreeHead(sth, pubkey); err != nil {
+		return fmt.Errorf("verifying checkpoint: %v", err)
+	}
+	log.Printf("Current checkpoint verified - Tree Size: %d Root Hash: %s\n", sth.Size, hex.EncodeToString(sth.Hash))
+
+	// If this is the very first snapshot within the monitor, save the snapshot
+	if first {
+		if err := verifySnapshot(sth, logInfoFile); err != nil {
+			return err
+		}
+	}
+
+	for {
+		// Check for root hash consistency
+		newSTH, err := mirroring.VerifyLogConsistency(rc.client, int64(sth.Size), sth.Hash)
+		if err != nil {
+			return fmt.Errorf("failed to verify log consistency: %v", err)
+		} else {
+			log.Printf("Root hash consistency verified - Tree Size: %d Root Hash: %s\n", newSTH.Size, hex.EncodeToString(newSTH.Hash))
+		}
+
+		// Append new, consistency-checked snapshot
+		if newSTH.Size != sth.Size {
+			s, err := newSTH.SignedNote.MarshalText()
+			if err != nil {
+				return fmt.Errorf("failed to marshal STH: %v", err)
+			}
+
+			// Open file to append new snapshot
+			file, err := os.OpenFile(logInfoFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+			if err != nil {
+				return fmt.Errorf("failed to open log file: %v", err)
+			}
+			defer file.Close()
```

Comment:
this is going to end up creating a lot of file handles that don't close since this loop runs for a long time.